backend/ethereum/channel/contractbackend.go

```diff
@@ -198,6 +197,81 @@ func (c *ContractBackend) ConfirmTransaction(ctx context.Context, tx *types.Tran
 	return receipt, nil
 }
 
+// confirmNTimes waits for a transaction to be included in `finalityDepth`
+// many consecutive blocks.
+// If a reorg happens that is deeper than `finalityDepth`, the behaviour is
+// undefined.
+// TODO: Think about adding a fast-path for reverts, since they probably do
+// need to wait for `finalityDepth` many blocks.
+func (c *ContractBackend) confirmNTimes(ctx context.Context, tx *types.Transaction, finalityDepth uint64) (*types.Receipt, error) {
+	// Set up header sub for future blocks.
+	heads := make(chan *types.Header, 10)
+	hsub, err := c.SubscribeNewHead(ctx, heads)
+	if err != nil {
+		err = cherrors.CheckIsChainNotReachableError(err)
+		return nil, errors.WithMessage(err, "subscribing to heads")
+	}
+	defer hsub.Unsubscribe()
+
+	// Wait to be included at least once. TxFinalityDepth starts at 1 anyway.
+	head, err := c.waitMined(ctx, tx)
+	if err != nil {
+		return nil, errors.WithMessage(err, "waiting for TX to be mined")
+	}
+	heads <- head
+
+	for {
+		select {
+		case head := <-heads:
+			// Poll the receipt of the TX.
+			receipt, err := c.ContractInterface.TransactionReceipt(ctx, tx.Hash())
+			if err != nil {
+				err = cherrors.CheckIsChainNotReachableError(err)
+				return nil, errors.WithMessage(err, "pulling receipt")
+			}
+			// TODO for the advanced version: In case the chain reorgs to a
+			// shorter canonical chain right here, it could incorrectly return.
+			// This case is not testable with the geth sim backend.
+			if receipt != nil && isFinal(receipt, head, finalityDepth) {
+				return receipt, nil
+			}
+			// TX was either is not included in the canonical chain anymore
+			// or is not yet final; wait for next head.
+		case err := <-hsub.Err():
+			err = cherrors.CheckIsChainNotReachableError(err)
+			return nil, errors.WithMessage(err, "header subscription")
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+}
+
+// waitMined waits for a TX to be mined and returns the latest head.
+func (c *ContractBackend) waitMined(ctx context.Context, tx *types.Transaction) (*types.Header, error) {
+	_, err := bind.WaitMined(ctx, c, tx)
+	if err != nil {
+		err = cherrors.CheckIsChainNotReachableError(err)
```

Comment:
Well, this code bloating was foreseeable. With more geth-dependent errors it will only get worse.  
I will try your approach for now, but on long term it would probably be better to do the wrapping elsewhere in cases where we dont need it.

Reply:
Looking at it: `CheckIsChainNotReachableError` also attaches the stack, so just deferring it would not work.

Reply:
OK, so you can see my comment as a general comment. Probably nothing that we can do about this within the scope of this PR.

Reply:
Created issue #158